os/net/bluetooth/bluetooth.c

```diff
@@ -1084,19 +1092,147 @@ int bt_conn_le_param_update(struct bt_conn *conn, const struct bt_le_conn_param
 
 int bt_conn_disconnect(struct bt_conn *conn, uint8_t reason)
 {
-	// TODO: need to implement
-	return 0;
+	struct bt_conn_s *conns = (struct bt_conn_s *)conn;
+
+	/* Disconnection is initiated by us, so auto connection shall be disabled.
+	 * Otherwise the passive scan would be enabled and we could send LE Create
+	 * Connection as soon as the remote starts advertising.
+	 */
+
+	if (conns->type == BT_CONN_TYPE_LE) {
+		bt_le_set_auto_conn(&conns->dst, NULL);
+	}
+
+	switch (conns->state) {
+	case BT_CONN_CONNECT_SCAN:
+		conns->err = reason;
+		bt_conn_set_state(conns, BT_CONN_DISCONNECTED);
+		bt_le_scan_update_internal(false);
+		return 0;
+	case BT_CONN_CONNECT_DIR_ADV:
+		conns->err = reason;
+		bt_conn_set_state(conns, BT_CONN_DISCONNECTED);
+		/* User should unref connection object when receiving
+		* error in connection callback.
+		*/
+		return bt_le_adv_stop();
+
+	case BT_CONN_CONNECT:
+		return bt_hci_connect_le_cancel(conns);
+
+	case BT_CONN_CONNECTED:
+		return bt_hci_disconnect(conns, reason);
+
+	case BT_CONN_DISCONNECT:
+		return 0;
+
+	case BT_CONN_DISCONNECTED:
+	default:
+		return -ENOTCONN;
+	}
 }
 
+
 struct bt_conn *bt_conn_create_le(const bt_addr_le_t *peer, const struct bt_le_conn_param *param)
 {
-	// TODO: need to implement
-	return NULL;
+	struct bt_conn_s *conn;
+
+	if (!bt_atomic_testbit(g_btdev.flags, BT_DEV_READY)) {
+		return NULL;
+	}
+
+	if (!bt_le_conn_params_valid(param)) {
+		return NULL;
+	}
+
+	if (bt_atomic_testbit(g_btdev.flags, BT_DEV_EXPLICIT_SCAN)) {
+		return NULL;
+	}
+
+	conn = bt_conn_lookup_addr_le_id(BT_ID_DEFAULT, peer);
+	if (conn) {
+		switch (conn->state) {
+		case BT_CONN_CONNECT_SCAN:
+			bt_conn_set_param_le(conn, param);
+			return (struct bt_conn *)conn;
+		case BT_CONN_CONNECT:
+		case BT_CONN_CONNECTED:
+			return (struct bt_conn *)conn;
+		default:
+			bt_conn_relref(conn);
+			return NULL;
+		}
+	}
+
+	conn = bt_conn_add_le(peer);
+	if (!conn) {
+		return NULL;
+	}
+
+	/* Only default identity supported for now */
+	conn->id = BT_ID_DEFAULT;
+
+	/* Set initial address - will be updated later if necessary. */
+	bt_addr_le_copy(&conn->le.resp_addr, peer);
+
+	bt_conn_set_param_le(conn, param);
+
+	bt_conn_set_state(conn, BT_CONN_CONNECT_SCAN);
+
+	bt_le_scan_update_internal(true);
+
+	return (struct bt_conn *)conn;
 }
 
 int bt_le_set_auto_conn(const bt_addr_le_t *addr, const struct bt_le_conn_param *param)
```

Comment:
Please add 'static' keyword.

Reply:
bt_le_set_auto_conn() is defined in include/tinyara/bluetooth/conn.h
So it could not be static